music21/stream/core.py

```diff
@@ -384,18 +384,26 @@ def coreGuardBeforeAddElement(self, element, *, checkRedundancy=True):
         Before adding an element, this method provides
         important checks to that element.
 
-        Used by both insert() and append()
+        Used by insert() and append() and Stream.__init__()
 
         Returns None or raises a StreamException
 
         >>> s = stream.Stream()
         >>> s.coreGuardBeforeAddElement(s)
         Traceback (most recent call last):
         music21.exceptions21.StreamException: this Stream cannot be contained within itself
+
+        >>> s.append(s.iter)
+        Traceback (most recent call last):
+        music21.exceptions21.StreamException: cannot insert StreamIterator into a Stream
+        Iterate over it instead (Users Guide chs. 6 and 26)
         '''
         # using id() here b/c we do not want to get __eq__ comparisons
         if element is self:  # cannot add this Stream into itself
             raise StreamException('this Stream cannot be contained within itself')
+        if 'StreamIterator' in element.classes:
+            raise StreamException('cannot insert StreamIterator into a Stream\n'
+            'Iterate over it instead (Users Guide chs. 6 and 26)')
```

Comment:
can we have a `try` except around this, because this is also the place presumably to check that the object IS a music21 object. Right here, we get two-for-one, because non-music21 objects will probably not have a `.classes` element.